src/foam/nanos/test/SmartTest.java

```diff
@@ -0,0 +1,145 @@
+package foam.nanos.test;
+
+import foam.core.X;
+import foam.core.EmptyX;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.Map;
+
+import static java.lang.System.err;
+
+public class SmartTest extends Test {
+  public static final String TEST_METHOD_SUFFIX = "Test";
+  public static final String CLEANUP_METHOD_SUFFIX = "Cleanup";
+  public static final String SETUP_METHOD_SUFFIX = "Setup";
+
+  private Method globalSetupMethod;
+  private Method globalCleanupMethod;
+  private Method preTestMethod;
+  private Method postTestMethod;
+  private final Map<String, TestChain> testChains = new HashMap<>();
+
+  protected X x_ = EmptyX.instance();
+
+  public X getX() {
+    return x_;
+  }
+
+  public void setX(X x) {
+    x_ = x;
+  }
+
+  @Override
+  public void runTest(X x) {
+    setX(x);
+    prepareTestChains();
+    runTestChains();
+  }
+
+  private void runTestChains() {
+    if ( invokeMethod(globalSetupMethod) ) {
+      for ( String test: testChains.keySet() ) {
+        TestChain testChain = testChains.get(test);
+        runTestChain(testChain);
+      }
+      invokeMethod(globalCleanupMethod);
+    }
+  }
+
+  private void runTestChain(TestChain testChain) {
```

Comment:
It would be nice if the current test name was recorded and used to prefix test(boolean,message) output so that you could easily test what test was producing test messages.